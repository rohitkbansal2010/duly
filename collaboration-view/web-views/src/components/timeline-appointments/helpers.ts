export const formatDateAppointment = (date: Date) =>
  new Intl.DateTimeFormat('en-US', {
    weekday: 'short',
    month: 'short',
    day: 'numeric',
    year: 'numeric',
  }).format(date);

export const formatTimeAppointment = (date: Date) =>
  new Intl.DateTimeFormat('en-US', {
    hour: 'numeric',
    minute: 'numeric',
    hour12: true,
  }).format(date);
