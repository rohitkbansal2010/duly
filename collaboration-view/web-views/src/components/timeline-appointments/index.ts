export { TimelineAppointments } from './timeline-appointments';
export type { TimelineAppointmentsPropsType } from './timeline-appointments';
