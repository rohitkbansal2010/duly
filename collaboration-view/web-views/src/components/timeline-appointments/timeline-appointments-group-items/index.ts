export { TimelineAppointmentsGroupItems } from './timeline-appointments-group-items';
export type { TimelineAppointmentsGroupItemsPropsType } from './timeline-appointments-group-items';
export type { TimelineAppointmentsGroupItemPropsType } from './timeline-appointments-group-item';
