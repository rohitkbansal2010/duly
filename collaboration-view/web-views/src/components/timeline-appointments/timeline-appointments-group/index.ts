export { TimelineApppointmentsGroup } from './timeline-appointments-group';
export type { TimelineApppointmentsGroupPropsType } from './timeline-appointments-group';
