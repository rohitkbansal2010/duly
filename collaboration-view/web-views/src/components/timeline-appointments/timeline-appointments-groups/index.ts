export { TimelineApppointmentsGroups } from './timeline-appointments-groups';
export { TimelineApppointmentsGroupsNoData } from './timeline-appointments-groups-no-data';
export type { TimelineApppointmentsGroupsPropsType } from './timeline-appointments-groups';
export type { TimelineApppointmentsGroupsNoDataPropsType } from './timeline-appointments-groups-no-data';
