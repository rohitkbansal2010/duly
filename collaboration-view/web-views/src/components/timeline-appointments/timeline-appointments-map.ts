import { PatientAppointmentAlias } from '@enums';
import { calendarTwoBlueIcon, calendarTwoMagentaIcon } from '@icons';

type TimelineAppointmentsMapType = {
  [key in PatientAppointmentAlias]: {
    title: string;
    colorTitle: string;
    icon: string;
    backgroundIcon: string;
    backgroundCount: string;
    skeletonCount?: number;
    showNoShowStatusCount: boolean,
    showNoShowCancelledCount: boolean,
  };
};

export const timelineAppointmentsMap: TimelineAppointmentsMapType = {
  [PatientAppointmentAlias.RECENT]: {
    title: 'Recent Appointments',
    colorTitle: 'LightBlue',
    icon: calendarTwoBlueIcon,
    backgroundIcon: 'Blue',
    backgroundCount: 'DarkBlue',
    skeletonCount: 6,
    showNoShowStatusCount: true,
    showNoShowCancelledCount: true,
  },
  [PatientAppointmentAlias.UPCOMING]: {
    title: 'Upcoming Appointments',
    colorTitle: 'Magenta',
    icon: calendarTwoMagentaIcon,
    backgroundIcon: 'Magenta',
    backgroundCount: 'Violet',
    skeletonCount: 6,
    showNoShowStatusCount: false,
    showNoShowCancelledCount: true,
  },
};
