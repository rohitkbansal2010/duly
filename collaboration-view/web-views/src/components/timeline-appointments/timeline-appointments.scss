@import "~@styles/colors.scss";
@import "~@styles/typography.scss";

.timelineAppointments {
  &Container {
    display: flex;
    flex-direction: column;
    width: 50%;

    &:nth-child(2n + 1) {
      margin-right: 0.7rem;
    }

    &:nth-child(2n) {
      margin-left: 0.7rem;
    }

    &Title {
      margin-bottom: 1.7rem;
      font-size: var(--font-size-24px-4px);
      // line-height: 1.625rem;
      @include nexa-heavy-highlight;

      &LightBlue {
        color: $dupage-light-blue;
      }

      &Magenta {
        color: $dupage-magenta;
      }
    }

    &Content {
      flex-grow: 1;
      min-height: 0;
    }
  }

  &Skeleton {
    display: flex;
    flex-direction: column;
    width: 50%;

    &Title {
      @include nexa-heavy-highlight;
      font-size: 1.5rem;
      line-height: 1.625rem;

      &LightBlue {
        color: $dupage-light-blue;
      }

      &Magenta {
        color: $dupage-magenta;
      }
    }

    &Item {
      position: relative;
      height: 6.25rem;
      margin-top: 1.25rem;
      margin-right: 1.5rem;
      background: $white;
      border-radius: 0.5rem;
    }
  }
}
