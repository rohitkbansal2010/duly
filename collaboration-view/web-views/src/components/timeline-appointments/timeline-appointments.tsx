import { useSelector } from 'react-redux';

import { PatientAppointmentAlias, AppointmentModuleWidgetsAlias } from '@enums';
import { useWidgetSkeleton } from '@hooks';
import { RootState } from '@redux/reducers';

import {
  TimelineApppointmentsGroups,
  TimelineApppointmentsGroupsNoData
} from './timeline-appointments-groups';
import { timelineAppointmentsMap } from './timeline-appointments-map';
import { TimelineAppointmentsSkeleton } from './timeline-appointments-skeleton';
import styles from './timeline-appointments.scss';

export type TimelineAppointmentsPropsType = {
  alias: PatientAppointmentAlias;
};

export const TimelineAppointments = ({ alias }: TimelineAppointmentsPropsType) => {
  const patientAppointments = useSelector(({ OVERVIEW_WIDGETS }: RootState) =>
    OVERVIEW_WIDGETS.patientAppointments);

  const isSkeletonShown = useWidgetSkeleton(AppointmentModuleWidgetsAlias.TIMELINE);

  const {
    title,
    colorTitle,
    icon,
    backgroundIcon,
    showNoShowStatusCount,
    showNoShowCancelledCount,
    ...rest
  } = timelineAppointmentsMap[alias];

  const isAppointmentGroups =
    patientAppointments &&
    patientAppointments[alias] &&
    patientAppointments[alias].length;
  const classesTimelineAppointmentsTitle = [ styles.timelineAppointmentsContainerTitle ]
    .concat(styles[`timelineAppointmentsContainerTitle${colorTitle}`])
    .join(' ');

  if (isSkeletonShown) {
    return (
      <TimelineAppointmentsSkeleton alias={alias} />
    );
  }

  return (
    <div className={styles.timelineAppointmentsContainer}>
      <h4 className={classesTimelineAppointmentsTitle}>{title}</h4>
      {isAppointmentGroups
        ? (
          <div className={styles.timelineAppointmentsContainerContent}>
            <TimelineApppointmentsGroups
              icon={icon}
              backgroundIcon={backgroundIcon}
              alias={alias}
              groups={patientAppointments[alias]}
              showNoShowStatusCount={showNoShowStatusCount}
              showNoShowCancelledCount={showNoShowCancelledCount}
              {...rest}
            />
          </div>
        )
        : (
          <TimelineApppointmentsGroupsNoData
            icon={icon}
            backgroundIcon={backgroundIcon}
            title={title}
            colorTitle={colorTitle}
          />
        )
      }
    </div>
  );
};
