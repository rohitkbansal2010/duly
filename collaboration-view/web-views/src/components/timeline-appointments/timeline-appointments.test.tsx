import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as redux from 'react-redux';

import {
  AppointmentModuleWidgetsAlias, PatientAppointmentAlias, PatientAppointmentStatus, PractitionerRole
} from '@enums';
import { PatientAppointments } from '@types';

import { TimelineAppointments, TimelineAppointmentsPropsType } from './timeline-appointments';
import { timelineAppointmentsMap } from './timeline-appointments-map';

const mockPatientAppointments: PatientAppointments = {
  [PatientAppointmentAlias.RECENT]: [
    {
      title: 'Physical Therapy',
      appointments: [
        {
          practitioner: {
            id: 'qwerty1',
            humanName: {
              familyName: 'Reyes',
              givenNames: [ 'Ana', 'Maria' ],
              prefixes: [ 'Dr.' ],
            },
            role: PractitionerRole.PCP,
          },
          isTelehealth: true,
          reason: 'Patient fell - Hip problims',
          status: PatientAppointmentStatus.COMPLETED,
          startTime: '2021-11-02T14:45:00Z',
          minutesDuration: 30,
        },
      ],
      nearestAppointmentPractitioner: {
        id: 'qwerty1',
        humanName: {
          familyName: 'Reyes',
          givenNames: [ 'Ana', 'Maria' ],
          prefixes: [ 'Dr.' ],
        },
        role: PractitionerRole.PCP,
      },
      nearestAppointmentDate: '2021-11-02',
      farthestAppointmentDate: '2021-07-15',
    },
  ],
  [PatientAppointmentAlias.UPCOMING]: [],
};
const mockProps: TimelineAppointmentsPropsType = { alias: PatientAppointmentAlias.RECENT };

const setUp = (props: TimelineAppointmentsPropsType) =>
  shallow(<TimelineAppointments {...props} />);

describe('TimelineAppointments', () => {
  let component: ReturnType<typeof shallow>;

  beforeEach(() => {
    jest
      .spyOn(redux, 'useSelector')
      .mockImplementation((selector: (state: unknown) => unknown) =>
        selector({ OVERVIEW_WIDGETS: { patientAppointments: mockPatientAppointments } }));
    jest
      .spyOn(redux, 'useDispatch')
      .mockReturnValue(jest.fn());

    component = setUp(mockProps);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should render the TimelineAppointments component', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render the title', () => {
    const expectedTitle = timelineAppointmentsMap[mockProps.alias].title;

    expect(component.find('h4').text()
      .includes(expectedTitle)).toBeTruthy();
  });

  it('should render the TimelineApppointmentsGroups component', () => {
    const {
      icon,
      backgroundIcon,
      backgroundCount,
      skeletonCount,
      showNoShowStatusCount,
      showNoShowCancelledCount,
    } = timelineAppointmentsMap[mockProps.alias];

    const expectedProps = {
      icon,
      backgroundIcon,
      alias: mockProps.alias,
      groups: mockPatientAppointments[mockProps.alias],
      backgroundCount,
      skeletonCount,
      showNoShowStatusCount,
      showNoShowCancelledCount,
    };

    const componentTimelineApppointmentsGroups = component.find('TimelineApppointmentsGroups');

    expect(componentTimelineApppointmentsGroups).toHaveLength(1);
    expect(componentTimelineApppointmentsGroups.props()).toEqual(expectedProps);
  });

  it('should render the TimelineApppointmentsGroupsNoData component if no patientAppointments by alias', () => {
    const {
      icon,
      backgroundIcon,
      title,
      colorTitle,
    } = timelineAppointmentsMap[mockProps.alias];

    const expectedProps = {
      icon,
      backgroundIcon,
      title,
      colorTitle,
    };

    jest
      .spyOn(redux, 'useSelector')
      .mockImplementation((selector: (state: unknown) => unknown) =>
        selector(
          { OVERVIEW_WIDGETS: { patientAppointments: { [PatientAppointmentAlias.RECENT]: [] } } }
        ));

    const component = setUp(mockProps);

    expect(toJson(component)).toMatchSnapshot();

    const componentTimelineApppointmentsGroupsNoDate = component.find('TimelineApppointmentsGroupsNoData');
    expect(componentTimelineApppointmentsGroupsNoDate).toHaveLength(1);
    expect(componentTimelineApppointmentsGroupsNoDate.props()).toEqual(expectedProps);
  });

  it('should render the TimelineAppointmentsSkeleton component while dispatch array of appointments', () => {
    jest
      .spyOn(redux, 'useSelector')
      .mockImplementation((selector: (state: unknown) => unknown) =>
        selector({
          OVERVIEW_WIDGETS: { patientAppointments: { [PatientAppointmentAlias.RECENT]: [] } },
          CURRENT_APPOINTMENT: {
            widgetsList: [
              { alias: AppointmentModuleWidgetsAlias.TIMELINE, isSkeletonShown: true },
            ],
          },
        }));

    const component = setUp(mockProps);

    expect(toJson(component)).toMatchSnapshot();
    expect(component.find('TimelineAppointmentsSkeleton')).toHaveLength(1);
    expect(component.find('h4')).toHaveLength(0);
    expect(component.find('TimelineApppointmentsGroups')).toHaveLength(0);
  });
});
