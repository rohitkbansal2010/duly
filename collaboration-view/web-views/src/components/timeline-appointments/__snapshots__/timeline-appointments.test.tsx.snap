// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimelineAppointments should render the TimelineAppointments component 1`] = `
<div
  className="timelineAppointmentsContainer"
>
  <h4
    className="timelineAppointmentsContainerTitle timelineAppointmentsContainerTitleLightBlue"
  >
    Recent Appointments
  </h4>
  <div
    className="timelineAppointmentsContainerContent"
  >
    <TimelineApppointmentsGroups
      alias="recentAppointments"
      backgroundCount="DarkBlue"
      backgroundIcon="Blue"
      groups={
        Array [
          Object {
            "appointments": Array [
              Object {
                "isTelehealth": true,
                "minutesDuration": 30,
                "practitioner": Object {
                  "humanName": Object {
                    "familyName": "Reyes",
                    "givenNames": Array [
                      "Ana",
                      "Maria",
                    ],
                    "prefixes": Array [
                      "Dr.",
                    ],
                  },
                  "id": "qwerty1",
                  "role": "PCP",
                },
                "reason": "Patient fell - Hip problims",
                "startTime": "2021-11-02T14:45:00Z",
                "status": "Completed",
              },
            ],
            "farthestAppointmentDate": "2021-07-15",
            "nearestAppointmentDate": "2021-11-02",
            "nearestAppointmentPractitioner": Object {
              "humanName": Object {
                "familyName": "Reyes",
                "givenNames": Array [
                  "Ana",
                  "Maria",
                ],
                "prefixes": Array [
                  "Dr.",
                ],
              },
              "id": "qwerty1",
              "role": "PCP",
            },
            "title": "Physical Therapy",
          },
        ]
      }
      icon="calendar-two-blue.svg"
      showNoShowCancelledCount={true}
      showNoShowStatusCount={true}
      skeletonCount={6}
    />
  </div>
</div>
`;

exports[`TimelineAppointments should render the TimelineAppointmentsSkeleton component while dispatch array of appointments 1`] = `
<TimelineAppointmentsSkeleton
  alias="recentAppointments"
/>
`;

exports[`TimelineAppointments should render the TimelineApppointmentsGroupsNoData component if no patientAppointments by alias 1`] = `
<div
  className="timelineAppointmentsContainer"
>
  <h4
    className="timelineAppointmentsContainerTitle timelineAppointmentsContainerTitleLightBlue"
  >
    Recent Appointments
  </h4>
  <TimelineApppointmentsGroupsNoData
    backgroundIcon="Blue"
    colorTitle="LightBlue"
    icon="calendar-two-blue.svg"
    title="Recent Appointments"
  />
</div>
`;
